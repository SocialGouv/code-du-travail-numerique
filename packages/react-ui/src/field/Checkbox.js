import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";

import { box, spacings } from "../theme.js";

export const InputCheckbox = ({ label, name, id, size, ...props }) => (
  <StyledLabel htmlFor={id} size={size}>
    <StyledCheckbox
      type="checkbox"
      name={name}
      id={id}
      size={size}
      {...props}
    />
    {label}
  </StyledLabel>
);

InputCheckbox.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  name: PropTypes.string.isRequired,
  size: PropTypes.string,
};

InputCheckbox.defaultProps = {
  size: "1.6rem",
};

const StyledLabel = styled.label`
  display: flex;
  font-size: ${(props) => props.size};
  cursor: pointer;
  user-select: none;
  -moz-user-select: none;
  -webkit-user-select: none;
  -ms-user-select: none;
`;

const StyledCheckbox = styled.input`
  position: relative;
  display: block;
  flex-shrink: 0;
  width: calc(${(props) => props.size} * 1.25);
  height: calc(${(props) => props.size} * 1.25);
  margin: calc(${(props) => props.size} / 5)
    calc(2 * ${(props) => props.size} / 3) 0 0;
  background: ${({ theme }) => theme.white};
  border: ${({ theme }) => box.border(theme.border)};
  border-radius: ${spacings.tiny};
  cursor: pointer;
  appearance: none;
  user-select: none;
  -moz-user-select: none;
  -webkit-user-select: none;
  -ms-user-select: none;
  &:before {
    position: absolute;
    top: 50%;
    left: 50%;
    display: none;
    width: calc(${(props) => props.size} / 3);
    height: calc(${(props) => props.size} / 1.7);
    border: ${({ theme }) => box.border(theme.white)};
    border-width: 0 calc(${(props) => props.size} / 10)
      calc(${(props) => props.size} / 10) 0;
    transform: translate(-50%, -60%) rotate(45deg);
    content: "";
  }
  &:checked {
    background-color: ${({ theme }) => theme.primary};
    &:before {
      display: block;
    }
  }
`;
