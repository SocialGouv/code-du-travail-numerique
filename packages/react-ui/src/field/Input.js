import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";

import { box, breakpoints, fonts, spacings } from "../theme.js";

export const Input = ({ icon: Icon, className, ...props }) => {
  return (
    <StyledWrapper className={className}>
      <StyledInput hasIcon={Boolean(Icon)} {...props} />
      {Icon && (
        <StyledIcon>
          <Icon />
        </StyledIcon>
      )}
    </StyledWrapper>
  );
};

Input.propTypes = {
  className: PropTypes.string,
  icon: PropTypes.elementType,
  name: PropTypes.string.isRequired,
};

Input.defaultProps = {
  icon: null,
};

const iconDateSvg = `<svg fill="none" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32"><path fill-rule="evenodd" clip-rule="evenodd" d="M20.947 7.601h1.601c1.472 0 2.521 1.198 2.521 2.669v2.306l-.001 10.392c0 1.472-1.049 2.669-2.521 2.669H8.669A2.672 2.672 0 016 22.967l.001-12.697a2.672 2.672 0 012.67-2.669h1.331V6h.999v1.601H20V6h.947v1.601zm1.6 17.036c.883 0 1.454-.786 1.454-1.67v-10.33h-17L7 22.967c0 .884.785 1.67 1.668 1.67h13.879zm-15.548-13h17.002l.001-1.367c0-.883-.57-1.601-1.453-1.601h-1.6v1.068H20V8.669h-9v1.068h-1V8.669H8.669C7.786 8.669 7 9.387 7 10.27v1.367zm9.677 6.277h-2.135a1.07 1.07 0 01-1.068-1.068v-2.135a1.07 1.07 0 011.068-1.068h2.135a1.07 1.07 0 011.068 1.068v2.135a1.07 1.07 0 01-1.068 1.068zm0-3.203h-2.135v2.135h2.136v-2.135zm3.203 3.203h2.135a1.07 1.07 0 001.068-1.068v-2.135a1.07 1.07 0 00-1.068-1.068H19.88a1.07 1.07 0 00-1.068 1.068v2.135a1.07 1.07 0 001.068 1.068zm0-3.203h2.135l.001 2.135H19.88v-2.135zm-8.54 8.541H9.204a1.07 1.07 0 01-1.068-1.068v-2.135a1.07 1.07 0 011.068-1.068h2.135a1.07 1.07 0 011.068 1.068v2.135a1.07 1.07 0 01-1.068 1.068zm0-3.203H9.204v2.135h2.136l-.001-2.135zm3.202 3.203h2.135a1.07 1.07 0 001.068-1.068v-2.135a1.07 1.07 0 00-1.068-1.068h-2.135a1.07 1.07 0 00-1.068 1.068v2.135a1.07 1.07 0 001.068 1.068zm0-3.203h2.135l.001 2.135h-2.136v-2.135zm7.473 3.203H19.88a1.07 1.07 0 01-1.068-1.068v-2.135a1.07 1.07 0 011.068-1.068h2.135a1.07 1.07 0 011.068 1.068v2.135a1.07 1.07 0 01-1.068 1.068zm0-3.203H19.88v2.135h2.136v-2.135z" fill="currentColor"/></svg>`;

const INPUT_HEIGHT = "5.4rem";

const StyledWrapper = styled.span`
  position: relative;
  display: inline-block;
  @media (max-width: ${breakpoints.mobile}) {
    width: 100%;
  }
`;

const StyledInput = styled.input`
  width: 100%;
  height: ${INPUT_HEIGHT};
  padding: 0 ${spacings.medium};
  padding-right: ${(props) => (props.hasIcon ? "5rem" : spacings.medium)};
  color: ${({ theme }) => theme.paragraph};
  font-weight: normal;
  font-size: ${fonts.sizes.default};
  font-family: "Open Sans", sans-serif;
  font-style: normal;
  line-height: inherit;
  text-align: ${(props) => (props.type === "number" ? "right" : "left")};
  background: ${({ theme }) => theme.white};
  border: 1px solid transparent;
  border-color: ${({ invalid, theme }) =>
    invalid ? theme.error : "transparent"};
  border-radius: ${box.borderRadius};
  box-shadow: ${({ theme }) => box.shadow.default(theme.secondary)};
  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    margin: 0;
    appearance: none;
  }
  &::-webkit-calendar-picker-indicator {
    display: block;
    width: ${spacings.large};
    height: ${spacings.large};
    margin-right: -${spacings.medium};
    color: rgba(0, 0, 0, 0);
    background-color: ${({ theme }) => theme.placeholder};
    cursor: pointer;
    opacity: 1;
    mask-image: url("data:image/svg+xml;,${encodeURIComponent(iconDateSvg)}");
  }
  &:invalid {
    border-color: ${({ theme }) => theme.error};
  }
  &::placeholder {
    color: ${({ theme }) => theme.placeholder};
  }
  &:focus {
    border-color: ${({ theme }) => theme.secondary};
  }
  &:focus::placeholder {
    color: transparent;
  }
  appearance: none;
  @media (max-width: ${breakpoints.mobile}) {
    padding: 0 ${spacings.small};
    padding-right: ${(props) => (props.hasIcon ? "5rem" : spacings.medium)};
  }
`;

const StyledIcon = styled.div`
  position: absolute;
  top: 1rem;
  right: ${spacings.small};
  width: 100%;
  max-width: ${spacings.large};
  height: 100%;
  max-height: ${spacings.large};
  color: ${({ theme }) => theme.placeholder};
`;
