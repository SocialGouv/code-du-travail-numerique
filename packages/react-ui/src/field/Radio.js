import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";

import { box } from "../theme.js";

export const InputRadio = ({ label, name, id, size, ...props }) => (
  <StyledDiv>
    <StyledRadio type="radio" name={name} id={id} size={size} {...props} />
    <StyledLabel htmlFor={id} size={size}>
      {label}
    </StyledLabel>
  </StyledDiv>
);

InputRadio.propTypes = {
  checked: PropTypes.bool,
  id: PropTypes.string.isRequired,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  size: PropTypes.string.isRequired,
  value: PropTypes.string,
};

InputRadio.defaultProps = {
  size: "1.6rem",
};

const StyledLabel = styled.label`
  display: flex;
  font-size: ${(props) => props.size};
  cursor: pointer;
`;

const StyledDiv = styled.div`
  display: flex;
  flex-direction: row;
`;

const StyledRadio = styled.input`
  position: relative;
  display: inline-block;
  flex-shrink: 0;
  width: calc(${(props) => props.size} * 1.25);
  height: calc(${(props) => props.size} * 1.25);
  margin: calc(${(props) => props.size} / 5)
    calc(2 * ${(props) => props.size} / 3) 0 0;
  padding: 0;
  line-height: inherit;
  background: ${({ theme }) => theme.white};
  border: ${({ theme }) => box.border(theme.border)};
  border-radius: 50%;
  box-shadow: none;
  cursor: pointer;
  appearance: none;
  &:before {
    position: absolute;
    top: 50%;
    left: 50%;
    display: none;
    width: calc(${(props) => props.size} / 1.5);
    height: calc(${(props) => props.size} / 1.5);
    background-color: ${({ theme }) => theme.primary};
    border-radius: 50%;
    transform: translate(-50%, -50%);
    content: "";
  }
  &:checked {
    background-color: ${({ theme }) => theme.white};
    border-color: ${({ theme }) => theme.primary};
    &:before {
      display: block;
    }
  }
`;
