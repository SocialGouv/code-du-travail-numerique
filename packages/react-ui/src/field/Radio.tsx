import React from "react";
import styled from "styled-components";

import { box } from "../theme.js";

type InputRadioProps = {
  label?: string;
  name?: string;
  id?: string;
  size?: number;
  onChange?: () => void;
};

export const InputRadio = ({
  label,
  name,
  id,
  size,
  onChange,
}: InputRadioProps) => (
  <StyledDiv>
    <StyledRadio
      type="radio"
      name={name}
      id={id}
      size={size}
      onChange={onChange}
    />
    <StyledLabel htmlFor={id} size={size}>
      {label}
    </StyledLabel>
  </StyledDiv>
);

InputRadio.defaultProps = {
  size: "1.6rem",
};

const StyledLabel = styled.label`
  display: flex;
  font-size: ${({ size }: InputRadioProps) => size};
  cursor: pointer;
`;

const StyledDiv = styled.div`
  display: flex;
  flex-direction: row;
`;

const StyledRadio = styled.input`
  position: relative;
  display: inline-block;
  flex-shrink: 0;
  width: calc(${(props) => props.size} * 1.25);
  height: calc(${(props) => props.size} * 1.25);
  margin: calc(${(props) => props.size} / 5)
    calc(2 * ${(props) => props.size} / 3) 0 0;
  padding: 0;
  line-height: inherit;
  background: ${({ theme }) => theme.white};
  border: ${({ theme }) => box.border(theme.border)};
  border-radius: 50%;
  box-shadow: none;
  cursor: pointer;
  appearance: none;
  &:before {
    position: absolute;
    top: 50%;
    left: 50%;
    display: none;
    width: calc(${(props) => props.size} / 1.5);
    height: calc(${(props) => props.size} / 1.5);
    background-color: ${({ theme }) => theme.primary};
    border-radius: 50%;
    transform: translate(-50%, -50%);
    content: "";
  }
  &:checked {
    background-color: ${({ theme }) => theme.white};
    border-color: ${({ theme }) => theme.primary};
    &:before {
      display: block;
    }
  }
`;
