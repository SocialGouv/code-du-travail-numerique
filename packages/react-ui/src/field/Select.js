import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";

import { ArrowDown } from "../icons/index.js";
import { animations, box, breakpoints, fonts, spacings } from "../theme";

export const Select = ({ children, disabled, className, ...props }) => (
  <StyledWrapper className={className}>
    <StyledSelect disabled={disabled} {...props}>
      {children}
    </StyledSelect>
    <StyledArrowDown aria-hidden="true" isDisabled={disabled}>
      <ArrowDown />
    </StyledArrowDown>
  </StyledWrapper>
);

Select.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
};

Select.defaultProps = {
  disabled: false,
};

const INPUT_HEIGHT = "5.4rem";

const StyledWrapper = styled.div`
  position: relative;
  display: inline-flex;
  flex: 1 1 auto;
  align-items: center;
  justify-content: flex-start;
  text-align: center;
  @media (max-width: ${breakpoints.mobile}) {
    width: 100%;
  }
`;

const StyledArrowDown = styled.div`
  position: absolute;
  top: ${spacings.base};
  right: ${spacings.base};
  width: 1.6rem;
  height: 1.6rem;
  color: ${({ isDisabled, theme }) =>
    isDisabled ? theme.placeholder : theme.primary};
  pointer-events: none;
`;

const StyledSelect = styled.select`
  width: 100%;
  height: ${INPUT_HEIGHT};
  padding: 0 ${spacings.medium} 0;
  padding-right: 5rem;
  color: ${({ theme }) => theme.paragraph};
  font-size: ${fonts.sizes.default};
  font-family: "Open Sans", sans-serif;
  vertical-align: middle;
  background-color: ${({ theme }) => theme.white};
  border: none;
  border-radius: ${box.borderRadius};
  box-shadow: ${({ theme }) => box.shadow.default(theme.secondary)};
  cursor: pointer;
  transition: border-color ${animations.transitionTiming} ease;
  appearance: none;
  /* Internet Explorer 11 specifics rules */
  &::-ms-expand {
    background-color: transparent;
    border: 0 transparent;
  }
  *::-ms-backdrop,
  & {
    padding-right: ${spacings.base};
  }
  &:invalid {
    border-color: ${({ theme }) => theme.error};
  }
  &:disabled {
    background-color: ${({ theme }) => theme.bgTertiary};
  }
`;
