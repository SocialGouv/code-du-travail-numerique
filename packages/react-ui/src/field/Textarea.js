import PropTypes from "prop-types";
import React, { useCallback, useEffect, useMemo, useState } from "react";
import styled from "styled-components";

import { ScreenReaderOnly } from "../ScreenReaderOnly";
import { Text } from "../Text";
import { box, breakpoints, fonts, spacings } from "../theme";
import { debounce } from "../utils/debounce";

export const Textarea = ({
  name,
  maxLength,
  onChange,
  showCounter,
  ...props
}) => {
  const [value, setValue] = useState("");
  const [a11yCounter, setA11Counter] = useState(maxLength);
  const hasOverflowed = maxLength ? maxLength - value.length <= 0 : false;

  const a11yCounterUpdate = useMemo(
    () => debounce(setA11Counter, 500),
    [setA11Counter]
  );

  useEffect(() => {
    return () => {
      if (!hasOverflowed) return;
      a11yCounterUpdate.cancel();
    };
  }, []);

  const changeHandler = useCallback(
    (event) => {
      setValue(event.target.value);
      if (onChange) {
        onChange(event);
      }
      a11yCounterUpdate(maxLength - event.target.value.length);
    },
    [a11yCounterUpdate]
  );
  return (
    <>
      <StyledTextarea
        name={name}
        onChange={changeHandler}
        maxLength={maxLength}
        {...props}
      />
      {maxLength && showCounter && (
        <div>
          <Text
            fontSize="tiny"
            variant={hasOverflowed ? "error" : "placeholder"}
          >
            {Math.max(0, maxLength - value.length)} caractères restants
          </Text>
          <ScreenReaderOnly role="status">
            {a11yCounter} caractères restants
          </ScreenReaderOnly>
        </div>
      )}
    </>
  );
};

Textarea.propTypes = {
  maxLength: PropTypes.number,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  showCounter: PropTypes.bool,
};

const StyledTextarea = styled.textarea`
  width: 100%;
  min-height: 12rem;
  padding: ${spacings.medium};
  color: ${({ theme }) => theme.paragraph};
  font-size: ${fonts.sizes.default};
  font-family: "Open Sans", sans-serif;
  line-height: inherit;
  background: ${({ theme }) => theme.white};
  border: 1px solid transparent;
  border-radius: ${box.borderRadius};
  box-shadow: ${({ theme }) => box.shadow.default(theme.secondary)};
  appearance: none;

  &::placeholder {
    color: ${({ theme }) => theme.placeholder};
  }

  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    margin: 0;
    appearance: none;
  }

  &:focus {
    border-color: ${({ theme }) => theme.secondary};
  }

  &:focus::placeholder {
    color: transparent;
  }

  @media (max-width: ${breakpoints.mobile}) {
    width: 100%;
  }
`;
