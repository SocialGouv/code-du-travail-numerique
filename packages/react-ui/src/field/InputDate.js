import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";

import { box, breakpoints, fonts, spacings } from "../theme.js";
import { INPUT_HEIGHT } from "./Input";

export const InputDate = ({ value, invalid, onChange, ...props }) => {
  const [date, setDate] = React.useState(value ?? "");
  const [isValid, setIsValid] = React.useState(true ?? !invalid);
  const [isFocus, setIsFocus] = React.useState(false);

  const onChangeDate = (event) => {
    const value = event.target.value;
    const inputType = event.nativeEvent.inputType;
    if (inputType === "deleteContentBackward") {
      setDate(value);
      return;
    }
    const onlyNumbers = value.replace(/\D/g, "");
    const lastValue = onlyNumbers[onlyNumbers.length - 1] ?? "";
    const newValue = value.slice(0, -1) + lastValue;
    if (newValue.length === 2 || newValue.length === 5) {
      setDate(newValue + "/");
    } else if (newValue.length <= 10) {
      setDate(newValue);
    }
    if (newValue.length <= 10) {
      const splitParts = newValue.split("/");
      const day = isNaN(Number(splitParts[0])) ? null : Number(splitParts[0]);
      const month = isNaN(Number(splitParts[1])) ? null : Number(splitParts[1]);
      const year = isNaN(Number(splitParts[2])) ? null : Number(splitParts[2]);
      const isYearValid = year && year >= 1900 && year <= 2100;
      const isMonthValid = month && month >= 1 && month <= 12;
      const isDayValid = day && day >= 1 && day <= 31;
      const isValidDate = /^\d{2}\/\d{2}\/\d{4}$/.test(newValue);
      const isValid = isYearValid && isMonthValid && isDayValid && isValidDate;
      setIsValid(isValid);
    }
    if (onChange) onChange(newValue);
  };

  const onChangeDatePicker = (event) => {
    const date = event.target.value;
    const realDate = date.split("-");
    const year = realDate[0] ?? "";
    const month = realDate[1] ?? "";
    const day = realDate[2] ?? "";
    const newValue = `${day}/${month}/${year}`;
    setDate(newValue);
  };

  const formatDate = () => {
    const splitParts = date.split("/");
    const day = splitParts[0] ?? "";
    const month = splitParts[1] ?? "";
    const year = splitParts[2] ?? "";
    return `${year}-${month}-${day}`;
  };

  return (
    <StyledWrapper isFocus={isFocus} isValid={isValid}>
      <StyledInput
        value={date}
        invalid={!isValid}
        onChange={onChangeDate}
        placeholder="jj/mm/aaaa"
        onFocus={() => setIsFocus(true)}
        onBlur={() => setIsFocus(false)}
        {...props}
      />
      <DatePickerInputDate
        aria-disabled="true"
        type="date"
        min="1900-01-01"
        max="2100-01-01"
        maxlength="11"
        value={formatDate()}
        onChange={onChangeDatePicker}
      />
    </StyledWrapper>
  );
};

const iconDateSvg = `<svg fill="none" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32"><path fill-rule="evenodd" clip-rule="evenodd" d="M20.947 7.601h1.601c1.472 0 2.521 1.198 2.521 2.669v2.306l-.001 10.392c0 1.472-1.049 2.669-2.521 2.669H8.669A2.672 2.672 0 016 22.967l.001-12.697a2.672 2.672 0 012.67-2.669h1.331V6h.999v1.601H20V6h.947v1.601zm1.6 17.036c.883 0 1.454-.786 1.454-1.67v-10.33h-17L7 22.967c0 .884.785 1.67 1.668 1.67h13.879zm-15.548-13h17.002l.001-1.367c0-.883-.57-1.601-1.453-1.601h-1.6v1.068H20V8.669h-9v1.068h-1V8.669H8.669C7.786 8.669 7 9.387 7 10.27v1.367zm9.677 6.277h-2.135a1.07 1.07 0 01-1.068-1.068v-2.135a1.07 1.07 0 011.068-1.068h2.135a1.07 1.07 0 011.068 1.068v2.135a1.07 1.07 0 01-1.068 1.068zm0-3.203h-2.135v2.135h2.136v-2.135zm3.203 3.203h2.135a1.07 1.07 0 001.068-1.068v-2.135a1.07 1.07 0 00-1.068-1.068H19.88a1.07 1.07 0 00-1.068 1.068v2.135a1.07 1.07 0 001.068 1.068zm0-3.203h2.135l.001 2.135H19.88v-2.135zm-8.54 8.541H9.204a1.07 1.07 0 01-1.068-1.068v-2.135a1.07 1.07 0 011.068-1.068h2.135a1.07 1.07 0 011.068 1.068v2.135a1.07 1.07 0 01-1.068 1.068zm0-3.203H9.204v2.135h2.136l-.001-2.135zm3.202 3.203h2.135a1.07 1.07 0 001.068-1.068v-2.135a1.07 1.07 0 00-1.068-1.068h-2.135a1.07 1.07 0 00-1.068 1.068v2.135a1.07 1.07 0 001.068 1.068zm0-3.203h2.135l.001 2.135h-2.136v-2.135zm7.473 3.203H19.88a1.07 1.07 0 01-1.068-1.068v-2.135a1.07 1.07 0 011.068-1.068h2.135a1.07 1.07 0 011.068 1.068v2.135a1.07 1.07 0 01-1.068 1.068zm0-3.203H19.88v2.135h2.136v-2.135z" fill="currentColor"/></svg>`;

InputDate.propTypes = {
  invalid: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  value: PropTypes.string,
};

const StyledWrapper = styled.div`
  display: flex;
  flex-direction: row;
  @media (max-width: ${breakpoints.mobile}) {
    width: 100%;
  }
  height: ${INPUT_HEIGHT};

  background: ${({ theme }) => theme.white};
  box-shadow: ${({ theme }) => box.shadow.default(theme.secondary)};

  border: 2px solid transparent;
  border-radius: ${box.borderRadius};
  border-color: ${({ isFocus, isValid, theme }) =>
    isFocus ? theme.secondary : !isValid ? theme.error : "transparent"};

  line-height: inherit;

  appearance: none;
  @media (max-width: ${breakpoints.mobile}) {
    padding: 0 ${spacings.small};
    padding-right: ${(props) => (props.hasIcon ? "5rem" : spacings.medium)};
  }
`;

const StyledInput = styled.input`
  padding: 0 ${spacings.medium};
  padding-right: ${(props) => (props.hasIcon ? "5rem" : spacings.medium)};
  color: ${({ theme }) => theme.paragraph};
  font-weight: normal;
  font-size: ${fonts.sizes.default};
  font-family: "Open Sans", sans-serif;
  font-style: normal;
  border-color: transparent;
  outline: none;
  border-radius: ${box.borderRadius};
  width: 160px;
  @media (max-width: ${breakpoints.mobile}) {
    width: 100%;
  }

  &::placeholder {
    color: ${({ theme }) => theme.placeholder};
  }
`;

const DatePickerInputDate = styled(StyledInput)`
  width: 40px;
  padding: 0;
  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    margin: 0;
    appearance: none;
  }

  &::-webkit-calendar-picker-indicator {
    display: block;
    width: ${spacings.large};
    height: ${spacings.large};
    color: rgba(0, 0, 0, 0);
    background-color: ${({ theme }) => theme.placeholder};
    cursor: pointer;
    opacity: 1;
    mask-image: url("data:image/svg+xml;,${encodeURIComponent(iconDateSvg)}");
    mask-repeat: no-repeat;
  }
`;
