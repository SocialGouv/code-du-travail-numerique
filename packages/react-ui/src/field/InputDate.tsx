import React from "react";
import styled from "styled-components";

import { box, breakpoints, fonts, spacings } from "../theme";
import { DefaultInputProps, INPUT_HEIGHT } from "./Input";

function formatValueToFr(value: string): string {
  const [year, month, days] = value.split("-");
  if (
    !year ||
    year.length !== 4 ||
    !month ||
    month.length !== 2 ||
    !days ||
    days.length !== 2
  )
    return value;
  return `${days}/${month}/${year}`;
}

function formatValueToEn(value: string): string {
  if (!value) return value;
  const [days, month, year] = value.split("/");
  return `${year}-${month}-${days}`;
}

export const InputDate = ({ value, onChange, invalid, ref, ...props }: any) => {
  const onChangeDate = (event: any) => {
    if (onChange) {
      onChange(formatValueToFr(event.target.value));
    }
  };

  return (
    <StyledWrapper isValid={!invalid}>
      <StyledInput
        {...props}
        onChange={onChangeDate}
        type="date"
        ref={ref}
        data-input="true"
        value={formatValueToEn(value)}
      />
    </StyledWrapper>
  );
};

InputDate.propTypes = DefaultInputProps;

const StyledWrapper = styled.div<{ isValid: Boolean }>`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  @media (max-width: ${breakpoints.mobile}) {
    width: 100%;
  }
  height: ${INPUT_HEIGHT};
  width: fit-content;

  background: ${({ theme }) => theme.white};
  box-shadow: ${({ theme }) => box.shadow.default(theme.secondary)};

  border-radius: ${box.borderRadius};
  border: 1px solid
    ${({ isFocus, isValid, theme }: any) =>
      isFocus ? theme.secondary : !isValid ? theme.error : "transparent"};

  line-height: inherit;

  appearance: none;
  @media (max-width: ${breakpoints.mobile}) {
    padding: 0 ${spacings.small};
    padding-right: ${(props: any) =>
      props.hasIcon ? "5rem" : spacings.medium};
  }
`;

const StyledInput = styled.input`
  padding: 0 ${spacings.medium};
  padding-right: ${(props: any) => (props.hasIcon ? "5rem" : spacings.medium)};
  color: ${({ theme }) => theme.paragraph};
  font-weight: normal;
  font-size: ${fonts.sizes.default};
  font-family: "Open Sans", sans-serif;
  font-style: normal;
  border-color: transparent;
  outline: none;
  border-radius: ${box.borderRadius};
  width: 192px;
  padding-right: 2px;
  color: ${({ theme, value }) => (value ? theme.paragraph : theme.placeholder)};
  @media (max-width: ${breakpoints.mobile}) {
    width: 100%;
  }
`;
