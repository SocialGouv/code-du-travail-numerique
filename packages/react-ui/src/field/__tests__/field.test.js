import "regenerator-runtime/runtime";

import { fireEvent, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";

import { InputCheckbox } from "../Checkbox";
import { Input } from "../Input";
import { InputDate } from "../InputDate";
import { InputRadio } from "../Radio";
import { Select } from "../Select";
import { Textarea } from "../Textarea";

describe("<InputCheckbox />", () => {
  it("should render checkbox field", () => {
    const { container } = render(
      <InputCheckbox label="option" name="checkbox" id="checkbox" />
    );
    expect(container).toMatchSnapshot();
  });
  it("should be able to update controlled checkbox", () => {
    const { container } = render(
      <InputCheckbox label="option" name="checkbox" id="checkbox" />
    );
    const testValue = "true";
    const input = container.querySelector("input");
    fireEvent.change(input, { target: { value: testValue } });
    expect(input.value).toEqual(testValue);
  });
});

describe("<Input />", () => {
  it("should render input field", () => {
    const { container } = render(<Input name="input" />);
    expect(container).toMatchSnapshot();
  });
  it("should call input onChange", () => {
    const handleChange = jest.fn();
    const { container } = render(
      <Input name="input" onChange={handleChange} />
    );
    const input = container.querySelector("input");
    fireEvent.change(input, { target: { value: "foo" } });
    expect(handleChange).toHaveBeenCalledTimes(1);
  });
});

describe("<InputDate />", () => {
  it("should render input date field", () => {
    const { container } = render(<InputDate name="input_date" />);
    expect(container).toMatchSnapshot();
  });
  it("should be able to update input date", () => {
    const { container } = render(<InputDate name="input_date" />);
    const testValue = "2020-01-01";
    const input = container.querySelector("input");
    fireEvent.change(input, { target: { value: testValue } });
    expect(input.value).toEqual(testValue);
  });
});

describe("<InputRadio />", () => {
  it("should render input radio field", () => {
    const { container } = render(
      <InputRadio label="option" name="radio" id="radio" />
    );
    expect(container).toMatchSnapshot();
  });
  it("should be able to update controlled input radio", () => {
    const { container } = render(
      <InputRadio label="option" name="radio" id="radio" />
    );
    const testValue = "true";
    const input = container.querySelector("input");
    fireEvent.change(input, { target: { value: testValue } });
    expect(input.value).toEqual(testValue);
  });
});

describe("<Select />", () => {
  it("should render select field", () => {
    const { container } = render(
      <Select>
        <option>...</option>
        <option>Option 1</option>
        <option>Option 2</option>
      </Select>
    );
    expect(container).toMatchSnapshot();
  });
  it("should call onChange", () => {
    const handleChange = jest.fn();
    const { container } = render(
      <Select onChange={handleChange}>
        <option>...</option>
        <option>Option 1</option>
        <option>Option 2</option>
      </Select>
    );
    const select = container.querySelector("select");
    fireEvent.change(select, { target: { value: "foo" } });
    expect(handleChange).toHaveBeenCalledTimes(1);
  });
});

describe("<Textarea />", () => {
  it("should render textarea", () => {
    const { container } = render(<Textarea name="textarea" />);
    expect(container).toMatchSnapshot();
  });
  it("should call textarea onChange", () => {
    const handleChange = jest.fn();
    const { container } = render(
      <Textarea name="input" onChange={handleChange} />
    );
    const textarea = container.querySelector("textarea");
    fireEvent.change(textarea, { target: { value: "foo" } });
    expect(handleChange).toHaveBeenCalledTimes(1);
  });
});
