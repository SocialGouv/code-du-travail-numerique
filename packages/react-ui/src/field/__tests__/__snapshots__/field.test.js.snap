// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputCheckbox /> should render checkbox field 1`] = `
<div>
  <label
    class="sc-aXZVg utbZE"
    for="checkbox"
  >
    <input
      class="sc-gEvEer jFblLw"
      id="checkbox"
      name="checkbox"
      type="checkbox"
    />
    option
  </label>
</div>
`;

exports[`<InputDate /> should render input date field 1`] = `
<div>
  <div
    class="sc-iGgWBj kZPIXv"
  >
    <input
      class="sc-gsFSXq igmAJA"
      data-input="true"
      name="input_date"
      type="date"
      value=""
    />
  </div>
</div>
`;
